### ISSUE TITLE
Gradient Dimension Mismatch

### ISSUE THREAD

**Issue by enweg** [FluxML/Zygote.jl#1192]
```Julia
using LinearAlgebra, Distributions, Flux, AdvancedHMC, Zygote

net = Chain(LSTM(1, 1), Dense(1, 1))
θ, re = Flux.destructure(net)

function loss(y, x, re, θ) 
    nn = re(θ)
    out = vec([nn(xx) for xx in x][end])
    et = eltype(y)
    return sum(logpdf.(Normal.(out, one(et)), y))
end

y = randn(Float32, 10)
x = [randn(Float32, 1, 10) for _ in 1:10]

loss(y, x, re, θ)

Zygote.gradient(θ -> loss(y, x, re, θ), θ)
```

The code above always throws the error

`
┌ Warning: Expected 16 params, got 34
└ @ Flux ~/.julia/packages/Flux/qAdFM/src/utils.jl:650
ERROR: DimensionMismatch("variable with size(x) == (16,) cannot have a gradient with size(dx) == (34,)")
Stacktrace:
 [1] (::ChainRulesCore.ProjectTo{AbstractArray, NamedTuple{(:element, :axes), Tuple{ChainRulesCore.ProjectTo{Float32, NamedTuple{(), Tuple{}}}, Tuple{Base.OneTo{Int64}}}}})(dx::Vector{Float32})
   @ ChainRulesCore ~/.julia/packages/ChainRulesCore/IzITE/src/projection.jl:227
 [2] _project
   @ ~/.julia/packages/Zygote/cCyLF/src/compiler/chainrules.jl:182 [inlined]
 [3] map(f::typeof(Zygote._project), t::Tuple{Vector{Float32}}, s::Tuple{Vector{Float32}})
   @ Base ./tuple.jl:246
 [4] gradient(f::Function, args::Vector{Float32})
   @ Zygote ~/.julia/packages/Zygote/cCyLF/src/compiler/interface.jl:77
 [5] top-level scope
   @ ~/Documents/RA Nalan/code/julia/BFlux_test.jl:18
`

Seems like something goes wrong. This only happens when one trains models like this. All goes fine with implicit gradients, but I need all parameters in a vector to train it in a Bayesian way. 

Same error is being thrown when replacing the LSTM layer with an RNN layer. Dense layers, on the other hand, work. So maybe something goes wrong during the restructuring of recurrent layers? Things also work when the LSTM layer is replaced by an RNN layer and ReverseDiff or ForwardDiff are being used. This makes me believe that is is caused somewhere in Zygote? I could not test the LSTM layers with ForwardDiff and ReverseDiff since their implementation is not compatible with those two AD frameworks.

**Comment by mcabbott** [comment]
This looks like one more `Flux.destructure` bug. The example here works if you `] add Flux#master`, using a completely new implementation (from Optimisers.jl). Other ways to stress-test the new one are welcomed.

**Comment by mcabbott** [comment]
Closing as I think this is fixed on Flux 0.13
